### PR TITLE
Manually reveal characters

### LINKED ISSUE #13: Manual hints
Allow the drawing player to manually reveal up to x letters of the word by clicking on them. Also highlight which words are currently shown to other players.

### PR DESCRIPTION
**Pull request by rknoll**
This allows drawing players to manually reveal characters to give
guessing players additional hints. Eligible players can click on
individual characters to reveal them which will then be highlighted
for them and revealed for guessing players.

Fixes #13

### CODE REVIEW

File: server/Game.js [hunk 1]
```diff
@@ -206,6 +206,24 @@ export default class {
...
   }
 
+  revealCharacter(player, index) {
+    // Only allow revealing characters when the player is allowed to draw.
+    if (!this.started || !this.canDraw(player)) return;
+    this.revealCharacterIndex(index);
+  }
+
+  revealCharacterIndex(index) {
+    if (!this.currentWordEncoded || this.currentWordEncoded.length <= index) return;
```
Comment (glennhartmann) [inline server/Game.js]: nit: 'index >= this.currentWordEncoded.length' is seems more intuitive, but maybe that's just me...

File: server/Game.js [hunk 2]
```diff
@@ -206,6 +206,24 @@ export default class {
     this.roundTimer = setTimeout(() => this.tickGameRound(), 1000);
   }
 
+  revealCharacter(player, index) {
```
Comment (glennhartmann) [inline server/Game.js:211]: the names 'revealCharacter' vs 'revealCharacterIndex' don't really describe the difference between these functions (the main thing being that 'revealCharacterIndex' is unconditional, while 'revealCharacter' does extra validation).

Maybe 'revealCharacterIfAllowed' vs 'revealCharacter'? And maybe a comment describing when to use each.

File: client/components/CurrentWord.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import { makeStyles } from '@material-ui/core/styles';
+import Box from '@material-ui/core/Box';
+import gameActions from '../store/actions/game';
+
+const useStyles = makeStyles((theme) => ({
+  word: {
+    fontSize: 14,
+    textAlign: 'center',
+    wordBreak: 'break-all',
+    [theme.breakpoints.up('sm')]: {
```
Comment (glennhartmann) [inline client/components/CurrentWord.js:13]: the touch targets might be a little small on mobile. It's also not very discoverable, currently. Both could be good follow-up CLs.

File: client/components/CurrentWord.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      outlineStyle: 'dashed',
+    },
+  },
+  characterRevealed: {
+    textDecoration: 'underline',
+    fontWeight: 'bold',
+  },
+}));
+
+const CurrentWord = ({ selectedWord, currentWord, revealCharacter }) => {
```
Comment (glennhartmann) [inline client/components/CurrentWord.js]: I'm confused about the difference between |selectedWord| and |currentWord|

**Review by glennhartmann — CHANGES_REQUESTED** [review]

File: server/Game.js (hunk 1 above)
Comment (rknoll) [reply server/Game.js]: Done

File: server/Game.js (hunk 2 above)
Comment (rknoll) [reply server/Game.js:211]: I'd prefer to keep calling it `revealCharacter` as it aligns with other APIs and message names throughout the code. Note that we have similar checks in other methods like `command()` that only allows players who can draw to submit commands. Let me know if you feel strongly :-)
Added some more comments to both methods but they are quite short anyway.

File: client/components/CurrentWord.js (hunk 3 above)
Comment (rknoll) [reply client/components/CurrentWord.js:13]: True, maybe a different approach would be better on mobile? Maybe open a dialog to select the character to reveal? That would give us more space.. Definitely a follow-up :-)

File: client/components/CurrentWord.js (hunk 4 above)
Comment (rknoll) [reply client/components/CurrentWord.js]: Yeah that was unfortunate naming, renamed it now to match the server side.
`currentWord` now contains the cleartext word if it's your turn to draw or if you've guessed it correctly
`currentWordEncoded` contains the encoded form, as in, `_ _ _ a _ _` etc.

**Review by glennhartmann — APPROVED** [review]

**Pull request merged**